### PR TITLE
Add support for a subset of pyOpenSSL 19.0.0.

### PR DESCRIPTION
**Pull request by haydenroche5**
Specifically, the subset needed to support the Python module ndg_httpsclient.

### CODE REVIEW

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]
And some additional comments on slack.

File: pyopenssl/19.0.0/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Building pyOpenSSL with wolfSSL
++ It's recommended that you do the following steps using a Python virtual environment. Create it with `python -m venv /path/to/new/virtual/environment` and activate it with `source /path/to/new/virtual/environment/bin/activate`.
++ Clone wolfssl-py with `git clone git@github.com:wolfSSL/wolfssl-py.git`. `cd wolfssl-py`.
++ Build and install wolfssl-py with `pip install .`.
++ Download pyOpenSSL 19.0.0 with `curl -O https://files.pythonhosted.org/packages/40/d0/8efd61531f338a89b4efa48fcf1972d870d2b67a7aea9dcf70783c8464dc/pyOpenSSL-19.0.0.tar.gz`.
++ Extract the code with `tar xvf pyOpenSSL-19.0.0.tar.gz`. `cd pyOpenSSL-19.0.0`.
++ Apply the pyopenssl-19.0.0.patch file with `patch -p1 < pyopenssl-19.0.0.patch` (assuming the patch file is in the pyopenssl-19.0.0 directory; adjust the path according to your situation). This patch adds wolfSSl support.
++ Install pyOpenSSL with `pip install .`. At this time, only portions of pyOpenSSL have been ported to work with wolfSSL. Specifically, we've ported the parts used by the module ndg_httpsclient. The next section covers running the ndg_httpsclient unit tests.
```
[the remark this answers is not in the document]
Comment (haydenroche5) [reply pyopenssl/19.0.0/README.md:8]: Lots of ways.

1. OpenSSL bindings aren't imported if using wolfSSL: https://github.com/wolfSSL/osp/pull/69/files#diff-2395ea83c6404179ca9d7c23bd069537bcefc6a1f0d62ab1fdd2818b959d6e21R92
2. Stepping through failing unit tests with the debugger.
3. Turning on debug logging from wolfSSL.
4. OpenSSL bindings aren't anywhere standard where Python would be able to find it on my system (i.e. it's not on PYTHONPATH).

File: pyopenssl/19.0.0/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# Building pyOpenSSL with wolfSSL
++ It's recommended that you do the following steps using a Python virtual environment. Create it with `python -m venv /path/to/new/virtual/environment` and activate it with `source /path/to/new/virtual/environment/bin/activate`.
++ Clone wolfssl-py with `git clone git@github.com:wolfSSL/wolfssl-py.git`. `cd wolfssl-py`.
++ Build and install wolfssl-py with `pip install .`.
```
[the remark this answers is not in the document]
Comment (haydenroche5) [reply pyopenssl/19.0.0/README.md:4]: Yeah, I can modify the wolfssl-py PR to add that to the instructions. Don't think it belongs in this OSP PR.

File: pyopenssl/19.0.0/pyopenssl-19.0.0.patch [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
++                # object when it's time to do garbage collection. With wolfSSL,
++                # when that X509_free call happens, it's possible that the X509
++                # object has already been freed by FreeX509 (an internal
++                # function). That function doesn't care what the reference count
++                # is and does the free unconditionally. This causes problems
++                # when the garbage collector comes along and tries to free the
++                # already freed X509. The solution for the wolfSSL case is to
++                # duplicate the object rather than fiddling with the ref count.
++                x509_copy = _lib.X509_dup(x509)
++                cert = X509._from_raw_x509_ptr(x509_copy)
```
Comment (julek-wolfssl) [inline pyopenssl/19.0.0/pyopenssl-19.0.0.patch:57]: All calls to `FreeX509` should be changed to `wolfSSL_X509_free`. Leaving it like this is just going to cause problems down the line. I've found the following calls. Please change all calls to `wolfSSL_X509_free`. Then this section should not need a wolfSSL exception.
```
$ grep -rP '[^_]FreeX509\(' --include=*.[ch]
wolfssl/internal.h:    WOLFSSL_LOCAL void FreeX509(WOLFSSL_X509*);
src/ssl.c:        FreeX509(&ssl->peerCert);
src/ssl.c:static void ExternalFreeX509(WOLFSSL_X509* x509)
src/ssl.c:                FreeX509(x509);
src/ssl.c:    ExternalFreeX509(x509);
src/internal.c:void FreeX509(WOLFSSL_X509* x509)
src/internal.c:    FreeX509(&ssl->peerCert);
src/internal.c:                    FreeX509(x509);
src/internal.c:            FreeX509(x509);
src/internal.c:                            FreeX509(&ssl->peerCert);
src/internal.c:                            FreeX509(&ssl->peerCert);
```

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]

File: pyopenssl/19.0.0/README.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+# Testing
++ Download ndg_httpsclient 0.5.1 with `curl -O https://files.pythonhosted.org/packages/b9/f8/8f49278581cb848fb710a362bfc3028262a82044167684fb64ad068dbf92/ndg_httpsclient-0.5.1.tar.gz`.
++ Extract the code with `tar xvf ndg_httpsclient-0.5.1.tar.gz`. `cd ndg_httpsclient-0.5.1`.
++ Patch ndg_httpsclient to update the test certs and keys with `patch -p1 < ndg_httpsclient-0.5.1.patch` (assuming the patch file is in the ndg_httpsclient-0.5.1 directory; adjust the path according to your situation). This patch doesn't modify the core ndg_httpsclient code; it just makes the test assets usable.
++ Start the test server in the background using the provided script. First, `cd ndg/httpsclient/test/`. Then, `./scripts/openssl_https_server.sh > server.log 2>&1 &`
++ `cd ../../..` to get back to the root directory.
++ Run the tests (all should pass):
+    + `python -m ndg.httpsclient.test.test_https`.
+    + `python -m ndg.httpsclient.test.test_urllib2`.
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply pyopenssl/19.0.0/README.md:18]: See next review.

File: pyopenssl/19.0.0/pyopenssl-19.0.0.patch (hunk 3 above)
Comment (julek-wolfssl) [reply pyopenssl/19.0.0/pyopenssl-19.0.0.patch:57]: Obviously only `ExternalFreeX509` should be calling `FreeX509` (which means that `FreeX509` could be a static function in `ssl.c`.

Comment (haydenroche5) [reply pyopenssl/19.0.0/pyopenssl-19.0.0.patch:57]: Makes sense to me. I'll create a wolfSSL PR today or Monday doing that (I'm off tomorrow).

**Comment by JacobBarthelmeh** [comment]
https://github.com/wolfSSL/osp/pull/150

**Pull request closed without merging**